File: code/mplayer-spot.py
#!/usr/bin/python2 -O

"""Mplayer Spot remembers the spot in which you last exited 
mplayer and resumes from that spot the next time you start 
watching the same movie."""

import sys
import os
import re
from subprocess import Popen, PIPE, STDOUT

### SOME config vars

# dot directory to save files in
mplayerspot_dot_dir = os.path.expanduser("~/.mplayer-spot")
spots_dir = os.path.join(mplayerspot_dot_dir, "spots")

# Smallest length of movie to save position in seconds.
# For instance, we will not save the position for files
# under 20 minutes.
minimum_save_length = 20 * 60 # 30 minutes

# Leftover length of movie to consider it fully watched in seconds.
# For example, we will consider a movie fully watched even if 
# we exit with 5 minutes left.
leftover_okay_length = 5 * 60 # 5 minutes


class read_mplayer_line:
    """This is just a class that produces an iterator for reading over
    lines of mplayer output.  A special class is needed for this because
    mplayer users ansi escape sequences to manipulate the cursor, and 
    it is taken care of in this method.  Those special sequences are
    just converted to newlines."""

    def __init__(self, mplayer_stdout):
        self.stdout = mplayer_stdout
        self.eof_reached = False

    def next(self):
        if self.eof_reached:
            raise StopIteration

        line = ''
        while True:
            read_char = self.stdout.read(1)
            # if we can't read anything then we have come to EOF
            if not read_char:
                self.eof_reached = True
                read_char = '\n'

            line += read_char
            #line = self._fix_cursor_moving_esc_seq(line)

            # we only continue to process if this is the end of a line
            if line[-4:] == '\x1b[J\x0d' or line[-1] == '\n':
                return line

    def _fix_cursor_moving_esc_seq(self, line):
        """Mplayer uses ansi escape sequences to move the cursor
        around.  This takes out thos escape sequences and just
        replaces them with a newline.  If there are no escape sequences,
        this just returns line."""
        if line[-4:] == '\x1b[J\x0d':
            return line[:-4] + '\n'
        else:
            return line

    def __iter__(self):
        return self

def get_filename(line):
    "Gets the filename of a movie file."
    if line.startswith("ID_FILENAME="):
        return line.split('=', 1)[1]
    else:
        return None

def get_length(line):
    "Gets the total length of the file."
    if line.startswith("ID_LENGTH="):
        return float(line.split('=')[1])

def is_close(allowed_difference, a, b):
    "Returns true if a and b are close together."
    return abs(a - b) <= allowed_difference

# regular expression to make sure we match the video and audio position
cur_pos_regex = re.compile("A: *(?P<audio_time>\d+\.\d+) V: *(?P<video_time>\d+\.\d+) A-V:")
def get_cur_pos(line):
    """Gets the current position in the movie file by using a regex
    (so that we can save the position and return to it later.)"""
    result = cur_pos_regex.match(line)

    if not result:
        return None

    audio_time = float(result.group("audio_time"))
    video_time = float(result.group("video_time"))
    assert is_close(1.0, audio_time, video_time)
    return video_time


def write_out_spot_files(spots_dir, filename, length, ending_pos):
    """For each file in filename[], writes out a file in spots_dir 
    named file with the correct ending_pos (as long as it should be kept)."""
    if (len(filename) != len(length) or len(length) != len(ending_pos)):
        return

    for i in range(len(filename)):
        save_path = os.path.join(spots_dir, os.path.basename(filename[i].strip()))

        if length[i] < minimum_save_length:
            print "not saving because length[i] (" + str(length[i]) + ") is too short."
            continue
        if is_close(leftover_okay_length, length[i], ending_pos[i]):
            print "not saving because ending_pos[i] (" + str(ending_pos[i]) + \
                    ") is close to length[i] (" + str(length[i]) + ")."
            if os.path.exists(save_path):
                print "deleting old save file because we have finished this file."
                os.remove(save_path)
            continue

        print "saving to path: " + save_path
        f = open(save_path, 'w')
        f.write(str(ending_pos[i]) + "\n")

def seek_to_correct_location(mplayer_stdin, spots_dir, filename):
        save_path = os.path.join(spots_dir, os.path.basename(filename.strip()))

        if os.path.exists(save_path):
            f = open(save_path, "r")
            leftoff = int(float(f.read().strip()))
            if leftoff > 21:
                leftoff -= 20
            leftoff_string = str(leftoff)
            print "writing to mplayer: seek " + leftoff_string + "\n"
            mplayer_stdin.write("seek " + leftoff_string + "\n")



def main():

    # try to open .mplayer-spot/ and .mplayer-spot/spots/.
    # If they doesn't exist, create them
    if not os.path.isdir(mplayerspot_dot_dir):
        os.mkdir(mplayerspot_dot_dir)
    if not os.path.isdir(spots_dir):
        os.mkdir(spots_dir)


    mplayer_process = Popen(["mplayer", "-identify", "-slave"] + sys.argv[1:], 
            stdout=PIPE, stderr=STDOUT, stdin=PIPE, bufsize=5)

    mplayer_stdout = mplayer_process.stdout
    mplayer_stderr = mplayer_process.stderr
    mplayer_stdin = mplayer_process.stdin

    cur_movie_index = -1
    filenames = []
    lengths = []
    ending_pos = []

    try:
        for line in read_mplayer_line(mplayer_stdout):
            sys.stdout.write(line)

            # if we read in a new filename, then we can update our arrays
            new_filename = get_filename(line)
            if new_filename != None:
                filenames.append(new_filename)
                seek_to_correct_location(mplayer_stdin, spots_dir, new_filename)
                lengths.append(None)
                ending_pos.append(None)
                cur_movie_index += 1

            if cur_movie_index < 0:
                continue

            if not lengths[cur_movie_index]:
                lengths[cur_movie_index] = get_length(line)

            tmp_cur_pos = get_cur_pos(line)
            if tmp_cur_pos:
                ending_pos[cur_movie_index] = tmp_cur_pos
    except KeyboardInterrupt:
        pass


    assert (len(filenames) == len(ending_pos) and len(ending_pos) == len(lengths))
    print "filenames are: " + str(filenames)
    print "Lengths are: " + str(lengths)
    print "exit positions are: " + str(ending_pos)

    if len(lengths) == 0 or len(ending_pos) == 0:
        print sys.argv[0] + ": ERROR! Could not play movie."
        sys.exit(1)

    write_out_spot_files(spots_dir, filenames, lengths, ending_pos)


if __name__ == '__main__':
    main()

File: code/mplayer-spot.hs
#!/usr/bin/env stack
-- stack --resolver=lts-4.0 runghc --package=async --package=filepath --package=directory --package=process --package=bytestring --package=text --package=conduit-extra --package=conduit-combinators

-- This gives us an experience similar to @ghci@.  Raw values (1 and
-- 10 below) that implement the 'Num' type class will be defaulted to 'Int'
-- (specified with the @default@ command below).  Raw values (all strings)
-- that implement the 'IsString' typeclass will be defaulted to Text (also
-- specified with the @default@ command below).
--
-- Without this extension turned on, ghc will produce errors like this:
-- @
--  code.hs:27:25:
--      No instance for (Data.String.IsString a0)
--            arising from the literal ‘"-alF"’
-- @
{-# LANGUAGE ExtendedDefaultRules #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeFamilies #-}
{-# OPTIONS_GHC -Wall #-}

-- Don't warn that things (like strings and numbers) are being defaulted to
-- certain types.  It's okay because this is just shell programming.
{-# OPTIONS_GHC -fno-warn-type-defaults #-}

-- import Prelude hiding (takeWhile)

import Conduit (iterMC)
import Control.Applicative ((<|>))
import Control.Concurrent.Async (Concurrently(..))
import Control.Concurrent.MVar (MVar, modifyMVar_, newMVar, tryReadMVar)
import Control.Exception (IOException, finally, try)
import Control.Monad (void)
import Control.Monad.IO.Class (liftIO)
import Data.Attoparsec.ByteString.Char8
    ( Parser, char, notInClass, parseOnly, rational, skipSpace, skipWhile
    , string, takeWhile1
    )
import Data.ByteString (ByteString)
import qualified Data.ByteString.Char8 as C8
import Data.Conduit
    ( (=$=), ($$), Conduit, Producer, Sink, Source, awaitForever, yield )
import Data.Conduit.Combinators (stderr, stdin, stdout)
import Data.Conduit.Process (streamingProcess, waitForStreamingProcess)
import Data.Monoid ((<>))
import Data.Streaming.Process (StreamingProcessHandle)
import Data.Text (unpack)
import qualified Data.Text as T
import Data.Text.Encoding (decodeUtf8)
import System.Directory (createDirectoryIfMissing, getHomeDirectory, removeFile)
import System.Environment (getArgs)
import System.Exit (exitWith)
import System.FilePath ((</>))
import System.IO (BufferMode(..), hSetBuffering)
import qualified System.IO as IO
import System.Process (proc)

-- Define the the types that should be defaulted to.  We can define one
-- type for string-like things, and one type for integer-like things.  It
-- doesn't matter what order they are in.
default (T.Text, Int)

-- | A config for our program.
data Config = Config { configMPlayerSpotRCDir :: FilePath -- ^ @~/.mplayer-spots@ dir
                     , configSpotsDir :: FilePath         -- ^ @~/.mplayer-spots/spots@ dir
                     , configIgnoreSeconds :: Float       -- how many seconds to ignore
                                                          -- in media before creating
                                                          -- spot file
                     }
    deriving Show

-- | Create a default 'Config'.
defaultConfig :: IO Config
defaultConfig = do
    homeDir <- getHomeDirectory
    let rcDir = homeDir </> ".mplayer-spot"
    let spotsDir = rcDir </> "spots"
    let ignoreSeconds = 180
    pure $ Config rcDir spotsDir ignoreSeconds

-- | Info about our media.
data MediaInfo = MediaInfo { mediaInfoLength :: Maybe Float          -- ^ length of the media
                           , mediaInfoFilename :: Maybe ByteString   -- ^ filename of the media
                           , mediaInfoCurPos :: Maybe Float          -- ^ current position
                           , mediaInfoAlreadySetOldLocation :: Bool  -- ^ whether we have already
                                                                     -- set the old location
                           }
    deriving Show

defaultMediaInfo :: MediaInfo
defaultMediaInfo = MediaInfo { mediaInfoLength = Nothing
                             , mediaInfoFilename = Nothing
                             , mediaInfoCurPos = Nothing
                             , mediaInfoAlreadySetOldLocation = False
                             }

-- | Datatype to hold mplayer's stdin, stdout, etc conduits.
data MPlayer = MPlayer { mplayerStdin :: Sink ByteString IO ()
                       , mplayerStdout :: Source IO ByteString
                       , mplayerStderr :: Source IO ByteString
                       , mplayerProcHandle :: StreamingProcessHandle
                       }

-- | Take in a bunch of arguments and use them to create the mplayer process.
createMPlayerProcess :: [String] -> IO MPlayer
createMPlayerProcess programArgs = do
    let mplayerArgs = ["-identify", "-slave"] <> programArgs
    (processStdin, processStdout, processStderr, processHandle) <-
        streamingProcess (proc "mplayer" mplayerArgs)
    return $! MPlayer processStdin processStdout processStderr processHandle

-- | Parser for a value prefixed by a bytestring.  Uses skipWhile to make it
-- faster.
genericParser :: forall a . ByteString -> Parser a -> ByteString -> Maybe a
genericParser str parser mplayerLine =
    either (const Nothing) Just $ parseOnly go mplayerLine
  where
    go :: Parser a
    go = do
        -- Skip until we find the first character of the prefix that we are looking for.
        skipWhile (/= C8.head str)
        -- Try to match the prefix. If it matches, run the parser.
        string str *> parser
            -- If it doesn't match, then strip the first character and recurse.
            <|> char (C8.head str) *> go

-- | Parser for the length of the of the media.
getLength :: ByteString -> Maybe Float
getLength = genericParser "ID_LENGTH=" rational

-- | Parser for the filename of the of the media.
getFilename :: ByteString -> Maybe ByteString
getFilename = genericParser "ID_FILENAME=" $ takeWhile1 (notInClass "\n")

-- | Parser for the current location in the media.
getCurPos :: ByteString -> Maybe Float
getCurPos = genericParser "A:" $ skipSpace *> rational

updateLength :: MVar MediaInfo -> ByteString -> IO ()
updateLength mediaInfoMVar mplayerLine = do
    maybeMediaInfo <- tryReadMVar mediaInfoMVar
    case maybeMediaInfo of
        Just (MediaInfo (Just _) _ _ _) -> pure ()
        _ ->
            case getLength mplayerLine of
                Nothing -> pure ()
                Just mediaLength ->
                    modifyMVar_ mediaInfoMVar $ updateMediaInfo mediaLength
  where
    updateMediaInfo :: Float -> MediaInfo -> IO MediaInfo
    updateMediaInfo mediaLength mediaInfo =
        pure mediaInfo { mediaInfoLength = Just mediaLength }

updateFilename :: MVar MediaInfo -> ByteString -> IO ()
updateFilename mediaInfoMVar mplayerLine = do
    maybeMediaInfo <- tryReadMVar mediaInfoMVar
    case maybeMediaInfo of
        Just (MediaInfo _ (Just _) _ _) -> pure ()
        _ ->
            case getFilename mplayerLine of
                Nothing -> pure ()
                Just mediaFilename ->
                    modifyMVar_ mediaInfoMVar $ updateMediaInfo mediaFilename
  where
    updateMediaInfo :: ByteString -> MediaInfo -> IO MediaInfo
    updateMediaInfo mediaFilename mediaInfo =
        pure mediaInfo { mediaInfoFilename = Just mediaFilename }

updateCurPos :: MVar MediaInfo -> ByteString -> IO ()
updateCurPos mediaInfoMVar mplayerLine =
    maybe (pure ()) (modifyMVar_ mediaInfoMVar . updateMediaInfo) $ getCurPos mplayerLine
  where
    updateMediaInfo :: Float -> MediaInfo -> IO MediaInfo
    updateMediaInfo mediaCurPos mediaInfo =
        pure mediaInfo { mediaInfoCurPos = Just mediaCurPos }

-- | Try to read 3 different things from the mplayer stdout:
--
--   - length of the media
--   - media filename
--   - current position in the media
processMPlayerStdout :: MVar MediaInfo -> ByteString -> IO ()
processMPlayerStdout mediaInfoMVar mplayerLine = do
    updateLength mediaInfoMVar mplayerLine
    updateFilename mediaInfoMVar mplayerLine
    updateCurPos mediaInfoMVar mplayerLine

-- | 'Conduit' that reads in key presses (from stdin), and yields mplayer
-- commands.
--
-- In order for this to work right when hooked up to stdin, stdin should be set
-- to 'NoBuffering'.
sendMPlayerCommands :: Conduit ByteString IO ByteString
sendMPlayerCommands = awaitForever go
  where
    go :: ByteString -> Conduit ByteString IO ByteString
    go inputChar
        | inputChar == "q"       = yield "quit\n"
        | inputChar == "p"       = yield "pause\n"
        | inputChar == " "       = yield "pause\n"
        | inputChar == "\ESC[D"  = yield "seek -10\n"
        | inputChar == "\ESC[C"  = yield "seek 10\n"
        | inputChar == "\ESC[B"  = yield "seek -60\n"
        | inputChar == "\ESC[A"  = yield "seek 60\n"
        | inputChar == "\ESC[6~" = yield "seek -600\n"
        | inputChar == "\ESC[5~" = yield "seek 600\n"
        | otherwise =
            liftIO $ print $ "got keypress " <> inputChar <> " but don't know what to do with it"

-- | Producer that tries to read the filename from the @'MVar' 'MediaInfo'@,
-- and if it succeeds, tries to open the spot file and read in the saved
-- position.  Produce it as a value.
setOldLocation :: Config -> MVar MediaInfo -> Producer IO ByteString
setOldLocation config@(Config _ spotsDir _) mediaInfoMVar = do
    maybeMediaInfo <- liftIO $ tryReadMVar mediaInfoMVar
    case maybeMediaInfo of
        Just (MediaInfo _ _ _ True) -> pure ()
        Just (MediaInfo _ (Just filename) _ False) -> do
            filecontents <- liftIO $ try $ readFile $ spotsDir </> unpack (decodeUtf8 filename)
            case filecontents of
                Right oldLocation -> do
                    yield $ "seek " <> C8.pack oldLocation <> "\n"
                Left (_ :: IOException) -> pure ()
            liftIO $ modifyMVar_ mediaInfoMVar $ \mediaInfo -> pure mediaInfo { mediaInfoAlreadySetOldLocation = True }
        _ -> setOldLocation config mediaInfoMVar

-- | Run the mplayer process while doing 3 things:
--
--  - Read mplayer's stdout, looking for the length, filename, and current
--    position of the media.
--  - Once the filename has been found, look for a spot file, and if it exists,
--    make mplayer seek to the location.
--  - Take keys from stdin and translate them to mplayer commands.  Since mplayer
--    is in slave mode, it can't take in commands like normal.
runMPlayerUpdateMediaInfo :: Config -> MPlayer -> MVar MediaInfo -> IO ()
runMPlayerUpdateMediaInfo config mplayer mediaInfoMVar = do
    -- set stdin to not be buffered so we only get one character at a time
    hSetBuffering IO.stdin NoBuffering

    -- a conduit for sending the old location to mplayer
    let oldLocationToMplayerStdin =
            setOldLocation config mediaInfoMVar $$ mplayerStdin mplayer

    -- Read keypresses from stdin, translate to mplayer commands, and pipe to mplayer.
    let stdinToMplayerStdin = stdin $$
                              sendMPlayerCommands =$=
                              mplayerStdin mplayer

    -- Process mplayer's stdout to find filename, length, and current position.
    -- Print mplayer's stdout to our stdout.
    let mplayerStdoutToStdout = mplayerStdout mplayer $$
                                iterMC (processMPlayerStdout mediaInfoMVar) =$=
                                stdout

    -- Print mplayer's stderr to our stderr.
    let mplayerStderrToStderr = mplayerStderr mplayer $$ stderr

    -- Handle for controlling the mplayer process.
    let mplayerHandle = mplayerProcHandle mplayer

    -- Run all conduits concurrently.
    runConcurrently $
        Concurrently mplayerStdoutToStdout *>
        Concurrently mplayerStderrToStderr *>
        Concurrently oldLocationToMplayerStdin *>
        Concurrently stdinToMplayerStdin *>
        Concurrently (waitForStreamingProcess mplayerHandle >>= exitWith)

-- | Create the .mplayer-spots/ and spots/ directories from the config file.
createMPlayerSpotsDir :: Config -> IO ()
createMPlayerSpotsDir (Config rcDir spotsDir _) = do
    createDirectoryIfMissing True rcDir
    createDirectoryIfMissing True spotsDir

-- | Write out a spot file to the spots directory if all the required fields
-- have been filled in the MediaInfo, and if our current position in the media
-- file is not too early or not too late.
writeSpotFile :: Config -> MVar MediaInfo -> IO ()
writeSpotFile (Config _ spotsDir ignoreLength) mediaInfoMVar = do
    maybeMediaInfo <- tryReadMVar mediaInfoMVar
    case maybeMediaInfo of
        Just (MediaInfo (Just mediaLength) (Just filename) (Just exitPos) _) -> do
            let spotFilename = spotsDir </> unpack (decodeUtf8 filename)
            writeSpotFile' mediaLength spotFilename exitPos
        _ -> putStrLn "When exiting, do not currently have all fields of media info, so cannot write out spot file."
  where
    -- | Check that the length is not too early or too late.  If it is not,
    -- then write out the spot file.
    writeSpotFile' :: Float -> FilePath -> Float -> IO ()
    writeSpotFile' mediaLength spotFilename exitPos
        | exitPos <= ignoreLength = do
            putStrLn $ "exit position is " <> show exitPos <> " seconds so not writing spot file (not far enough)"
        | exitPos >= (mediaLength - ignoreLength) = do
            putStrLn $ "exit position is " <> show exitPos <> " seconds so not writing spot file (too close to end)"
            removeOldSpotFile spotFilename
        | otherwise =
            writeFloatToFile spotFilename exitPos

    -- | Write a Float to a FilePath.
    writeFloatToFile :: FilePath -> Float -> IO ()
    writeFloatToFile spotFilename exitPos = do
        putStrLn $ "writing to file: " <> spotFilename <> " (" <> show exitPos <> ")"
        writeFile spotFilename $ show exitPos

    -- | Remove a file and ignore errors (like if the file doesn't exist).
    removeOldSpotFile :: FilePath -> IO ()
    removeOldSpotFile =
        void . (try :: IO () -> IO (Either IOException ())) . removeFile

main :: IO ()
main = do
    -- read in program arguments
    programArgs <- getArgs

    -- create the config we will be using
    config <- defaultConfig

    -- create the .mplayer-spots directory if it doesn't exist
    createMPlayerSpotsDir config

    -- create the mplayer process
    mplayerProcess <- createMPlayerProcess programArgs

    -- create the MediaInfo MVar we will be using to do concurrent stuff
    mediaInfoMVar <- newMVar defaultMediaInfo

    finally (runMPlayerUpdateMediaInfo config mplayerProcess mediaInfoMVar) $
        -- write the spot file after exiting
        writeSpotFile config mediaInfoMVar
